### ISSUE TITLE
UDP scans with bandwidth (-B) option incorrectly compute probe size and send at incorrect rate

### ISSUE THREAD

**Issue by jhart-r7** [zmap/zmap#499]
(Historically related to #488)

I've noticed several issues with UDP scans that attempt to constrain the amount of bandwidth used during a scan, and it all seems to do with an improper computation of the probe size which is necessary to compute the proper packet rate to stay within the specified bandwidth option (`-B`).

For any UDP study that specifies a non-empty probe (`--probe-args:text:<text>`, `--probe-args:file:<some file>`, or `--probe-args:template:<some template>`), the size of the probe is always calculated as 84 bytes and ignores the UDP payload.

For example:

```
zmap -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=text:ffffffffff -B 1M 
zmap -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=file:/any/file -B 1M 
zmap -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:/any/file -B 1M
```

All result in using a probe size of 84:

```
Feb 05 10:29:16.895 [DEBUG] send: using bandwidth 1000000 bits/s for 84 byte probe, rate set to 1488 pkt/s
```

For the file and text option, it should compute the correct size of the overall probe by including the size of the text/file.  For templates it is a little harder because the template options used in the template are what ultimately dictate the size of the probe, and these are computed on a per-target basis.  In this case it may make sense to compute the rate based on the maximum size of the evaluated template by tracking the explicit size of the template options when possible, and when a field is src/target IP/port/etc dependent, pick a sane upper bound for each field.

**Comment by zakird** [comment]
@jhart-r7 can you confirm that this is fixed in master now that #499 is merged?

**Comment by jhart-r7** [comment]
Sure -- I think you mean #506

**Comment by zakird** [comment]
whoops! Yep, #506.

**Comment by jhart-r7** [comment]
For text probe types, it still looks to incorrectly set the size to 84:

```
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=text:ffff -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"

Mar 28 12:30:50.697 [DEBUG] send: using bandwidth 1000000 bits/s for 84 byte probe, rate set to 1488 pkt/s
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=text:ffffffff -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"

Mar 28 12:30:55.677 [DEBUG] send: using bandwidth 1000000 bits/s for 84 byte probe, rate set to 1488 pkt/s
```

For file probe types, it now looks correct in that the size of the probe increases directly inline with the file:

```
$  echo -n thisprobeis22byteslong > probe.txt                                                                                                                                  
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=file:probe.txt -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 28 12:34:51.416 [DEBUG] send: using bandwidth 1000000 bits/s for 88 byte probe, rate set to 1420 pkt/s
$  echo -n thisprobeis32byteslong1234567890 > probe.txt                                                                                                                         
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=file:probe.txt -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 28 12:35:27.844 [DEBUG] send: using bandwidth 1000000 bits/s for 98 byte probe, rate set to 1275 pkt/s
```

For templates, the size of the probe seems to scale relative to the non-template options used in the template:

```
$  echo -en '${RAND_ALPHA=16}\nFoo:bar' > test.tpl 
[ jhart@LAX-MBP-1576 (03/28/18 12:54:29) ~/rapid7/zmap  ]                                                                                                                                                                                   
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 28 12:54:34.688 [DEBUG] send: using bandwidth 1000000 bits/s for 90 byte probe, rate set to 1388 pkt/s
$  echo -en '${RAND_ALPHA=32}\nFoo:bar' > test.tpl                                                                                                                                
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 28 12:54:40.504 [DEBUG] send: using bandwidth 1000000 bits/s for 90 byte probe, rate set to 1388 pkt/s
$  echo -en '${RAND_ALPHA=32}\nFoo:barbaf' > test.tpl                                                                                                                             
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 28 12:54:47.906 [DEBUG] send: using bandwidth 1000000 bits/s for 93 byte probe, rate set to 1344 pkt/s
```

**Comment by zakird** [comment]
Re the first one, is this actually incorrect given that the minimum frame size is 84 bytes? If both of those sizes are < 84 bytes, it would be correct to mark them as 84 bytes, no?

**Comment by jhart-r7** [comment]
Oh right, correct.  I had that in my head when I tested file and template, but forgot to double check that for text.  It now seems to grow relative to the size of the provided text:

```
$   ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=text:fffffffffffffffffffff -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"

Mar 28 13:07:58.393 [DEBUG] send: using bandwidth 1000000 bits/s for 87 byte probe, rate set to 1436 pkt/s
$   ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=text:ffffffffffffffffffffffffffffffffffffff -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"

Mar 28 13:09:25.647 [DEBUG] send: using bandwidth 1000000 bits/s for 104 byte probe, rate set to 1201 pkt/s
```

**Comment by zakird** [comment]
Alright, so if I understand right the only issue is that the length of the templated values isn't being calculated correctly? Is that what you're seeing? Are templated values always a static length? Can we just run the templating once with a fake IP in `global_init` and use that? Or is this more complicated and/or intractable?

**Comment by jhart-r7** [comment]
@zakird yes, the length of templated values seems to be calculated incorrectly.  It seems as if it is coming from the length of the string that indicates the value.

Find that after 18-bytes we've gone beyond the minimum frame size:

```
$  echo -n '123456789012345678' > test.tpl                                                                                                                                        
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 29 09:32:04.780 [DEBUG] send: using bandwidth 1000000 bits/s for 84 byte probe, rate set to 1488 pkt/s
$  echo -n '1234567890123456789' > test.tpl                                                                                                                                         
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 29 09:32:09.563 [DEBUG] send: using bandwidth 1000000 bits/s for 85 byte probe, rate set to 1470 pkt/s
```

Now add 8 random alpha to the template.  We expect a length of 92 now (84 + 8), but get 99 instead:

```
$  echo -n '123456789012345678${RAND_ALPHA=8}' > test.tpl                                                                                                                           
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 29 09:32:44.931 [DEBUG] send: using bandwidth 1000000 bits/s for 99 byte probe, rate set to 1262 pkt/s


Adding another 8 random alpha, we see that the length grows 15 instead of 8:

```
$  echo -n '123456789012345678${RAND_ALPHA=8}${RAND_ALPHA=8}' > test.tpl                                                                                                       
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 29 09:33:13.001 [DEBUG] send: using bandwidth 1000000 bits/s for 114 byte probe, rate set to 1096 pkt/s
```

In this case, 15 is the length of `${RANDOM_ALPHA=8}`.

**Comment by jhart-r7** [comment]
No, the replaced tokens vary according to what is specified:

See that when we go from 8 to 9 random, the length doesn't change:

```
$  echo -n '123456789012345678${RAND_ALPHA=8}' > test.tpl                                                                                                            
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 29 10:26:55.209 [DEBUG] send: using bandwidth 1000000 bits/s for 99 byte probe, rate set to 1262 pkt/s
$  echo -n '123456789012345678${RAND_ALPHA=7}' > test.tpl                                                                                                                          
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 29 10:27:03.680 [DEBUG] send: using bandwidth 1000000 bits/s for 99 byte probe, rate set to 1262 pkt/s
```

See that when we change the length of the token, the size of probe changes:

```
$  echo -n '123456789012345678${RAND_ALPHA=10}' > test.tpl                                                                                                                         
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 29 10:27:10.623 [DEBUG] send: using bandwidth 1000000 bits/s for 100 byte probe, rate set to 1250 pkt/s
$  echo -n '123456789012345678${RAND_ALPHA=123}' > test.tpl                                                                                                                        
$  ./src/zmap -n 1 -p 5060 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data -v5 --probe-args=template:test.tpl -B 1M -G 00:00:00:00:00:00 2>&1 |grep "using band"
Mar 29 10:27:21.771 [DEBUG] send: using bandwidth 1000000 bits/s for 101 byte probe, rate set to 1237 pkt/s
```

As you can see, by adding 2 more characters to the token, the length of the probe increases by 2.

**Comment by zakird** [comment]
Sure, but it *will* always be 9 if specified that way? There aren't tokens that are sometimes 8 and sometimes 9 depending on the server they're getting sent to?

**Comment by jhart-r7** [comment]
I don't know.  I haven't examined everything that is sent out.  I am merely observing the calculated size that is displayed at startup, comparing it to what I expect and seeing how it changes as I use different template tokens.

**Comment by dadrian** [comment]
Templated fields can vary in size depending on the field type and destination. The fixed length field-types are:

```
SADDR_N: Source IP address in network byte order
DADDR_N: Destination IP address in network byte order
SPORT_N: UDP source port in netowrk byte order
DPORT_N: UDP destination port in network byte order
RAND_BYTE: Random bytes from 0-255
RAND_DIGIT: Random digits from 0-9
RAND_ALPHA: Random mixed-case letters (a-z)
RAND_ALPHANUM: Random mixed-case letters (a-z) and numbers
```

The variable length fields are

```
SADDR: Source IP address in dotted-quad format
DADDR: Destination IP address in dotted-quad format
SPORT: UDP source port in ascii format
DPORT: UDP destination port in ascii format
```

So, aside from whatever bug is screwing with UDP in general, we also can't rely on bandwidth calculations being consistent throughout the lifetime of the program. Alternatively, we could just do bandwidth calculations using the max size of each of those fields (5 for ports, 15 for dotted-quads).

**Comment by zakird** [comment]
Don't we allocate a buffer size in the beginning at least for how big the UDP packet is going to be? Can we use that to estimate?

**Comment by dadrian** [comment]
Yes, we allocate based on the max size of each field.

**Comment by dadrian** [comment]
This is now fixed as of #669 and 04e6c51
